### DISCUSSION TITLE
Autocomplete for python.

### DISCUSSION THREAD

**Discussion by ubaldot** [w0rp/ale#4261] [Q&A - Ask for help with problems]
I cannot make the autocomplete to work with Python, and I need some hint. :)
<c-x><c-o> won’t do anything. 
It looks like I did my homework, and  I also connected ale to the omnifunc in my .vimrc. 

set omnifunc=ale#completion#OmniFunc

I am using these linters now: 'pyright','jedils', 'flake8','pylsp','pycodestyle' but the autocomplete does not seem to work, not even with the classic numpy. :(

```

 Current Filetype: python
Available Linters: ['bandit', 'cspell', 'flake8', 'flakehell', 'jedils', 'mypy', 'prospector', 'pycodestyle', 'pydocstyle', 'pyflakes', 'pylama', 'pylint', 'pylsp', 'pyre', 'pyright', 'unimport', 'vulture']
  Enabled Linters: ['flake8', 'jedils', 'pycodestyle', 'pylsp', 'pyright']
  Ignored Linters: []
 Suggested Fixers: 
  'add_blank_lines_for_python_control_statements' - Add blank lines before control statements.
  'autoflake' - Fix flake issues with autoflake.
  'autoimport' - Fix import issues with autoimport.
  'autopep8' - Fix PEP8 issues with autopep8.
  'black' - Fix PEP8 issues with black.
  'isort' - Sort Python imports with isort.
  'pyflyby' - Tidy Python imports with pyflyby.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'reorder-python-imports' - Sort Python imports with reorder-python-imports.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
  'yapf' - Fix Python files with yapf.
 Linter Variables:

let g:ale_python_auto_pipenv = 1
let g:ale_python_auto_poetry = 0
let g:ale_python_black_options = '--line-length=89'
let g:ale_python_flake8_auto_pipenv = 0
let g:ale_python_flake8_auto_poetry = 0
let g:ale_python_flake8_change_directory = 'project'
let g:ale_python_flake8_executable = 'flake8'
let g:ale_python_flake8_options = '--ignore=E501, W503'
let g:ale_python_flake8_use_global = 0
let g:ale_python_jedils_auto_pipenv = 0
let g:ale_python_jedils_executable = 'jedi-language-server'
let g:ale_python_jedils_use_global = 0
let g:ale_python_pycodestyle_auto_pipenv = 0
let g:ale_python_pycodestyle_auto_poetry = 0
let g:ale_python_pycodestyle_executable = 'pycodestyle'
let g:ale_python_pycodestyle_options = ''
let g:ale_python_pycodestyle_use_global = 0
let g:ale_python_pylsp_auto_pipenv = 0
let g:ale_python_pylsp_auto_poetry = 0
let g:ale_python_pylsp_config = {}
let g:ale_python_pylsp_executable = 'pylsp'
let g:ale_python_pylsp_options = ''
let g:ale_python_pylsp_use_global = 0
let g:ale_python_pyright_config = {'pyright': {'executionEnvironments': {'root': 'C:/VAS/tools/python/model_validation', 'extraPaths': ['C:/VAS/tools/python/model_validation/dymoval'], 'pythonVersion': '3.10', 'pythonPlatform': 'Windows'}}}
let g:ale_python_pyright_executable = 'pyright-langserver'
 Global Variables:

let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = 100
let g:ale_completion_enabled = 1
let g:ale_completion_max_suggestions = 50
let g:ale_disable_lsp = 0
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%linter% says %s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 1
let g:ale_fixers = {'c': ['clang-format', 'remove_trailing_lines', 'trim_whitespace'], 'cpp': ['clang-format'], 'python': ['remove_trailing_lines', 'trim_whitespace', 'autoflake', 'black']}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'normal'
let g:ale_linter_aliases = {}
let g:ale_linters = {'c': ['clangd', 'cppcheck', 'gcc'], 'python': ['pyright', 'jedils', 'flake8', 'pylsp', 'pycodestyle']}
let g:ale_linters_explicit = 0
let g:ale_linters_ignore = {}
let g:ale_list_vertical = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '%linter% says %s'
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = v:null
let g:ale_open_list = 0
let g:ale_pattern_options = v:null
let g:ale_pattern_options_enabled = v:null
let g:ale_root = {}
let g:ale_set_balloons = 1
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'
let g:ale_sign_warning = '--'
let g:ale_sign_highlight_linenrs = 0
let g:ale_statusline_format = v:null
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 0
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(executable check - success) flake8
(finished - exit code 0) 'cmd /s/c "cd /d C:\VAS\tools\python\model_validation\scripts && flake8 --version"'

<<<OUTPUT STARTS>>>
3.9.2 (mccabe: 0.6.1, pycodestyle: 2.7.0, pyflakes: 2.3.1) CPython 3.9.13 on
Windows
<<<OUTPUT ENDS>>>

(executable check - success) jedi-language-server
(executable check - success) pycodestyle
(finished - exit code 0) 'cmd /s/c "pycodestyle  - < C:\Users\yt75534\AppData\Local\Temp\VQYDCB0.tmp\dymoval_tutorial.py"'

<<<NO OUTPUT RETURNED>>>

(executable check - success) pylsp
(executable check - success) pyright-langserver
(finished - exit code 2) 'cmd /s/c "cd /d C:\VAS\tools\python\model_validation\scripts && flake8 --ignore=E501, W503 --format=default --stdin-display-name C:\VAS\tools\python\model_validation\scripts\dymoval_tutorial.py - < C:\Users\yt75534\AppData\Local\Temp\VRRDF70.tmp\dymoval_tutorial.py"'

<<<NO OUTPUT RETURNED>>>

(started) 'cmd /s/c "jedi-language-server"'
(started) 'cmd /s/c "pylsp"'
```

**Comment by hsanson** [comment]
You have jedils, pyls, and pyright language servers enabled. I would recommend you choose and use only one of them.  I use jedils on my projects and auto-completion works fine.

Note that I use the asyncomplete integration and do not use ALE's built in autocompletion:  `:h ale-asyncomplete-integration`

**Comment by ubaldot** [comment]
Strange. I tried to isolate each LSP, none worked. 
I also tried to comment out these lines in mi `.vimrc`

```
let g:ale_completion_enabled = 1
let g:ale_completion_autoimport = 1
set omnifunc=ale#completion#OmniFunc
```

Nothing happened. The` :h ale-asyncomplete-integration` says that the autocomplete won't work if `paste `if active but it is quite cryptic and I am not sure what does it mean. 
Any hints? :)

**Reply by ubaldot** [reply]
just found this from the docs:

ALE Automatic completion implementation replaces |completeopt| before opening
the omnicomplete menu with <C-x><C-o>. In some versions of Vim, the value set
for the option will not be respected. If you experience issues with Vim
automatically inserting text while you type, set the following option in
vimrc, and your issues should go away. >
  set completeopt=menu,menuone,preview,noselect,noinsert


I’ll try one of these days, but any additional hint is more than welcome 😃

**Reply by hsanson** [reply]
Check if ALE is being able to find the root of your project. Without a root the LSP servers won't be started:

```
:echo ale#python#FindProjectRoot(bufnr("%"))
```

If this command returns empty then ALE is unable to find a root and the LSPs won't start.  Chek `autoload/ale/python.vim` to learn what files ALE looks for to determine root paths (e.g. setup.cfg, tox.ini, etc.).

If you are using vim (not neovim) add this to your configuration:

```
call ch_logfile(expand('/tmp/chlogfile.log'), 'w')
```

Then open some python file and see the logs in /tmp/chlogfile.log. It may provide more details on what is going on.

**Reply by ubaldot** [reply]
I finally found some time for doing the tests.

`:echo ale#python#FindProjectRoot(bufnr("%"))`

return the correct path.
I also included the following to my .vimrc

`call ch_logfile(expand('/tmp/chlogfile.log'), 'w')`

and I cannot find anything bad there. If needed, I can post it as well.

**Reply by ubaldot** [reply]
**UPDATE:** I re-installed Anaconda and now **almost everything** works fine: linting, fixing and autocomplete are in place.  
I have no idea why, but so far so good! :) 
The only non-working thing is the recognition of the custom package from pyright. 

Small suggestion: given that ALE also works as LSP client, wouldn't be better to add a section for it? 
At the moment both LSP and linters are under the set of "linters". It would be great to have "available linters, fixers, lsp" and "enabled linters, fixers and lsp".

What do you think? :)

**Comment by ubaldot** [comment]
**FIXED**
Re-installing `Anaconda `made the job. 
I have both `pyright `and `pylsp `working, now I only have to choose which one to use 😄 
The only thing left if that upon the pressure of `.` I don't see the pop-up with all the completion options immediately, but I have to type something first and then go back. 

For example, assume that I `import mypackage as mpk`. 
Then, when I type `mpk.` nothing happens, but if I type e.g. `mpk.s` then a pop-up with all the suggestions starting with letter `s` appear and then if I delete the letter `s`, thus obtaining `mpk.` I can see all the possible complete options in mpk. 
Is there a way to show all the options straight away upon pressure of` .`?

**Reply by hsanson** [reply]
Using this configuration and editing a simple hello world python program I get the completion menu appear after pressing ".":

```
let g:ale_completion_enabled = 1

set runtimepath=$VIMRUNTIME
set runtimepath^=$HOME/Projects/vim/ale
set packpath=
" set nocompatible

"let g:ale_virtualtext_cursor = 1
let g:ale_linters = {'python': ['jedils', 'pylint', 'flake8']}
```

**Reply by ubaldot** [reply]
Hum, interesting. But don't you have something like this: https://ibb.co/BqKkxsz that then get resolved as soon as you start typing, see this: https://ibb.co/kHXVSVC?

**Reply by hsanson** [reply]
This is how it works for me using the minimal vimrc I posted above. Note this uses ALE's auto-completion:

![ale-autocomplete](https://user-images.githubusercontent.com/900716/182144334-077de4f1-d3f8-4ca1-86b5-473647da1598.gif)


And this is how it works for me using my usual configuration. Here I use asyncomplete to do the auto completion:

![tty](https://user-images.githubusercontent.com/900716/182144425-5fd69e15-c812-4ed8-bfc4-03ea24259135.gif)

Auto-completion menu appears automatically as I type on both examples. I do not start omnicompletion pressing Ctrl-X Ctrl-O.
